### ISSUE TITLE
Allow to add both messages and errors on the same page using WC_Admin_Settings

### ISSUE THREAD

**Issue by webdados** [woothemes/woocommerce#53397]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

After saving settings and making some API calls to external services, I need to show 2 or 3 notices depending on the result of the calls. Some might be OK and some might be errors.

If I only use `WC_Admin_Settings::add_message` and `WC_Admin_Settings::add_errror`, I can show all the 2 or 3 notices, but I just found out that if I use both, only the errors are shown, because there is a condition that if errors exist, only those are shown: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/admin/class-wc-admin-settings.php#L125

Any reasoning behind for this?

I made a small change from this:
```
public static function show_messages() {
	if ( count( self::$errors ) > 0 ) {
		foreach ( self::$errors as $error ) {
			echo '<div id="message" class="error inline"><p><strong>' . esc_html( $error ) . '</strong></p></div>';
		}
	} elseif ( count( self::$messages ) > 0 ) {
		foreach ( self::$messages as $message ) {
			echo '<div id="message" class="updated inline"><p><strong>' . esc_html( $message ) . '</strong></p></div>';
		}
	}
}
```
To this:
```
public static function show_messages() {
	if ( count( self::$errors ) > 0 ) {
		foreach ( self::$errors as $error ) {
			echo '<div id="message" class="error inline"><p><strong>' . esc_html( $error ) . '</strong></p></div>';
		}
	}
	if ( count( self::$messages ) > 0 ) {
		foreach ( self::$messages as $message ) {
			echo '<div id="message" class="updated inline"><p><strong>' . esc_html( $message ) . '</strong></p></div>';
		}
	}
}
```
And I can show the three notices (errors + OK) without issues and I don't see any side-effects on doing so.
![Image](https://github.com/user-attachments/assets/7ee23fcd-6c6b-4765-b398-ac1d09d184d0)

### Expected behavior

Be able to add both errors and notices to the same admin page.

If more that one error or one message is allowed, there's no reason to not allow showing one or more of each.

### Actual behavior

If an error is added with `WC_Admin_Settings::add_errror` any message added with `WC_Admin_Settings::add_message` is not shown.

### Steps to reproduce

Add one notice of each and only the error is shown

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.5.0-beta.1
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.8.2
Log Directory Writable: ✔
WP Version: ❌ 6.8-alpha-59459 - There is a newer version of WordPress available (6.8-alpha-59479)
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: –
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.3.0
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.7.1
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

acf-field: 38
acf-field-group: 8
amn_smtp: 1
atividade: 4
attachment: 6284
aw_workflow: 2
blocks: 8
chefpanda_market: 3
doctor: 4
easy-pricing-table: 1
featured_item: 8
frm_form_actions: 1
frm_styles: 1
homey_reservation: 2
ibwp_mp_popup: 1
listing: 1
mb_payment_request: 1
medical-agreements: 3
medical-booking: 10
medical-consultation: 6
medical-exam: 2
medical-specialty: 4
memberpressproduct: 2
nav_menu_item: 25
oksofasstore: 3
page: 79
person: 5
post: 25
private_post: 4
product: 7967
product_variation: 136
revision: 358
shop_coupon: 23
shop_order_placehold: 991
shop_subscription: 10
store: 1
szbdzones: 1
was: 1
wc_appointment: 1
wc_conditional_fee: 1
wc_order_status: 9
wc-donation: 2
wc-donation-report: 6
wcdp_payment: 4
wd_client_contract: 1
wd_support_ticket: 1
webdados_client_serv: 12
webdados_service: 8
wfacp_checkout: 1
wffn_ty: 1
wishlist: 1
wp_block: 1
wp_global_styles: 5
wp_navigation: 1
wp_template: 1
wp_template_part: 1
wpbdp_listing: 1
wpcf7_contact_form: 6
wpforms: 1
wps_subscriptions: 1
wpsc_ticket_thread: 3

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins ###

Multibanco, MB WAY, Credit card, Payshop and Cofidis Pay (IfthenPay) for WooCommerce: by PT Woo Plugins (by Webdados) – 9.6.0
WooCommerce Beta Tester: by WooCommerce – 2.3.2
WooCommerce: by Automattic – 9.5.0-beta.1 (update to version 9.5.0-beta.2 is available)
WordPress Beta Tester: by WordPress Upgrade/Install Team – 3.6.1
WP Mail SMTP: by WP Mail SMTP – 4.2.0

### Inactive Plugins  ###

Irrelevant

### Dropin Plugins () ###

db.php: db.php

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: appointment (appointment)
external (external)
grouped (grouped)
mwb_booking (mwb_booking)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 2 MB

### WC Pages ###

Shop base: #14857 - /loja-pt/
Cart: #17882 - /cart-blocks/ -  Contains the woocommerce/cart block
Checkout: #17884 - /checkout-blocks-2/ -  Contains the woocommerce/checkout block
My account: #16787 - /my-account-2/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront Child
Version: 20230130.01
Author URL: 
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 4.6.0
Parent Theme Author URL: https://woocommerce.com/
Theme type: Classic theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings
use-wp-horizon
add-to-cart-with-options-stepper-layout
beta-tester-slotfill-examples

Daily Cron: ✔ Next scheduled: 2024-11-13 15:19:07 +00:00
Options: ✔
Notes: 185
Onboarding: completed

### Action Scheduler ###

Complete: 270
Oldest: 2024-11-06 13:09:12 +0000
Newest: 2024-12-03 18:11:01 +0000

Failed: 2,342
Oldest: 2023-02-01 17:31:32 +0000
Newest: 2024-12-02 13:01:50 +0000

Pending: 3
Oldest: 2024-12-04 10:39:59 +0000
Newest: 2024-12-25 12:46:49 +0000


### Status report information ###

Generated at: 2024-12-03 18:36:12 +00:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by timmy5685** [comment]
Hi @webdados - thanks for the report.

I know this has been open a while, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an designer's eyes on this before we make a call, so I've flagged it for design team feedback. Someone will take a proper look at this!

**Comment by keoshi** [comment]
Hey, Marco @webdados!

Thanks for such a clear report (and for your patience, this one's been open a while). You've identified something that actually kicked off a wider conversation on our side about how settings saves should give feedback.

Before we recommend anything, I'd love to understand your exact flow a bit better. A full screenshot or quick gif would help a lot:

- What does the settings screen look like when this happens?
- What are the external calls doing: validating or registering credentials with the gateway, something like that?
- Are these results really statuses of durable things (is this key valid, is this service connected), or one-off actions?
- How many results usually: two or three, or can it be a lot?
- When one fails, does the merchant need to fix something, or is it more of a "try again later"?

Here's where our thinking is heading, and I'm curious what you make of it: when one save kicks off several results, a notice per result gets noisy fast and notices disappear, so they stop being useful right when someone might want to check back.

What we'd lean towards instead: confirm the save itself once (a single success message) and show each result's status next to the thing it belongs to. For example, each service could show "Connected" or "Needs attention → Retry", something more permanent but, more importantly, more contextual. Take the Connectors UI as an example:

<img width="2220" height="1342" alt="Image" src="https://github.com/user-attachments/assets/f0d3e3c3-8d22-4fce-84b8-26a64f004c41" />

Do you think the same logic and UI pattern could be applied to your use case? If you can share a screenshot of the flow, that'd really help us understand the limitations of the current thinking.

Thanks again, Marco.

**Comment by webdados** [comment]
Hi,

1) This is what the settings screen looks like before saving:

<img width="1205" height="862" alt="Image" src="https://github.com/user-attachments/assets/0a3a9073-694c-415a-b7fd-ea72e2955e1a" />

2) The first time these are saved, we connect to the payment gateway API to "let them know" our callback URL for payment notifications for each of the payment methods. (We can also force it to run after the first time)

I forced it so you can see what the screen looks like after calling the API

<img width="881" height="756" alt="Image" src="https://github.com/user-attachments/assets/07e36af8-797d-4006-b603-bbfc5ca4cf8b" />

3) Mostly one-off actions (unless you change the gateway keys - each customer can have several, to use on different websites, and they'll show up on the select fields)

4) 3 at most for now - that's the number of payment methods we have on this particular screen. It can grow, but never too much, as we try to have each payment method as a single WooCommerce payment method.

5) It's a "try again later" thing most of the time.

Yes, we can show the "current" status next to each payment method, and we'll probably do that, which doesn't replace the fact that we should give immediate feedback, and that's exactly what notices were built for. We are not abusing the notices system 😂

I really appreciate the extent you went to in this internal discussion, but in fact I still think there's a bug, and a simple one to fix. My use case was just an example.

The underlying issue is this:

> **Expected behavior**
> Be able to add both errors and notices to the same admin page.
> If more that one error or one message is allowed, there's no reason to not allow showing one or more of each.
> **Actual behavior**
> If an error is added with WC_Admin_Settings::add_errror any message added with WC_Admin_Settings::add_message is not shown.

This can be solved by replacing `elseif` with `if`, as I mentioned when opening the issue.

**Comment by webdados** [comment]
For reference: the issue still exists [as of 11.0.0-rc.1](https://github.com/woocommerce/woocommerce/blob/11.0.0-rc.1/plugins/woocommerce/includes/admin/class-wc-admin-settings.php#L164)

**Comment by keoshi** [comment]
Thanks again for all the details, Marco. Appreciate it.

> This can be solved by replacing `elseif` with `if`, as I mentioned when opening the issue.

No doubt. But not knowing the genesis of the original code, I can only speculate that the `elseif` after `count( self::$errors ) > 0` served a purpose: to only show errors until those can be cleared out and replaced by a success message. In other words, I think it's trying to be conservative in getting users' attention to what needs immediate action, before ever communicating the favorable outcome.

My intention with the conversation above was to 1) understand where you were coming from and what specifics made you run into that and 2) give you an indication of a broader pattern we're considering for newer UIs.

In the meanwhile, I'll try to get an engineering team on this.

> Yes, we can show the "current" status next to each payment method, and we'll probably do that, which doesn't replace the fact that we should give immediate feedback

My point is that they don't have to be two separate mechanisms. If the status is updated whenever the field changes, that would also be a form of immediate feedback. Potentially even more contextual and immediate.

Again, not sure if possible in your case and that's why I asked. We are looking at moving settings screens to [DataForms](https://wordpress.github.io/gutenberg/?path=/docs/dataviews-dataform--docs) which would give us a whole palette of benefits, from field validation, async calls, derived data, and more, not to mention a nicer UI. :)

Edit: check this post: [Unifying extension settings in WooCommerce](https://developer.woocommerce.com/2026/07/08/unifying-extension-settings-in-woocommerce/) and [Call for testing: Settings UI support for WooCommerce extensions](https://developer.woocommerce.com/2026/07/08/settings-ui/).
